### PR TITLE
Add wrapped object factory test

### PR DESCRIPTION
**Pull request by brodybits**
Similar to the example at: https://nodejs.org/api/addons.html#addons_factory_of_wrapped_objects

### CODE REVIEW

File: test/cpp/wrappedobjectfactory.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+class MyObject : public ObjectWrap {
+ public:
+  static NAN_MODULE_INIT(Init) {
+    v8::Local<v8::FunctionTemplate> tpl = Nan::New<v8::FunctionTemplate>(New);
+    tpl->SetClassName(Nan::New("MyObject").ToLocalChecked());
+    tpl->InstanceTemplate()->SetInternalFieldCount(1);
+
+    SetPrototypeMethod(tpl, "getValue", GetValue);
+
+    constructor.Reset(tpl->GetFunction());
```
Comment (kkoopa) [inline test/cpp/wrappedobjectfactory.cpp]: `GetFunction(tpl).ToLocalChecked()`

File: test/cpp/wrappedobjectfactory.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  static NAN_METHOD(NewInstance) {
+    v8::Local<v8::Function> cons = Nan::New(constructor);
+    double value = info[0]->IsUndefined() ? 0 : info[0]->NumberValue();
```
Comment (kkoopa) [inline test/cpp/wrappedobjectfactory.cpp]: `To<double>(info[0]).FromJust()`

File: test/cpp/wrappedobjectfactory.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    v8::Local<v8::Value> argv[1] = {Nan::New(value)};
+    info.GetReturnValue().Set(cons->NewInstance(1, argv));
```
Comment (kkoopa) [inline test/cpp/wrappedobjectfactory.cpp]: `NewInstance(const, 1, argv).ToLocalChecked()`

File: test/cpp/wrappedobjectfactory.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+ private:
+  explicit MyObject(double value = 0) : value_(value) {}
+  ~MyObject() {}
+
+  static NAN_METHOD(New) {
+    if (info.IsConstructCall()) {
+      double value = info[0]->IsUndefined() ? 0 : info[0]->NumberValue();
```
Comment (kkoopa) [inline test/cpp/wrappedobjectfactory.cpp]: `To<double>(info[0]).FromJust()`

File: test/cpp/wrappedobjectfactory.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      MyObject * obj = new MyObject(value);
+      obj->Wrap(info.This());
+      info.GetReturnValue().Set(info.This());
+    } else {
+      const int argc = 1;
+      v8::Local<v8::Value> argv[argc] = {info[0]};
+      v8::Local<v8::Function> cons = Nan::New(constructor);
+      info.GetReturnValue().Set(cons->NewInstance(argc, argv));
```
Comment (kkoopa) [inline test/cpp/wrappedobjectfactory.cpp]: `NewInstance(const, 1, argv).ToLocalChecked()`

File: test/cpp/wrappedobjectfactory.cpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  double value_;
+};
+
+Persistent<v8::Function> MyObject::constructor;
+
+NAN_MODULE_INIT(Init) {
+  MyObject::Init(target);
+  Set(target
+    , New<v8::String>("newFactoryObjectInstance").ToLocalChecked()
+    , New<v8::FunctionTemplate>(MyObject::NewInstance)->GetFunction()
```
Comment (kkoopa) [inline test/cpp/wrappedobjectfactory.cpp]: `GetFunction(New<v8::FunctionTemplate>(MyObject::NewInstance)).ToLocalChecked()`

**Comment by kkoopa** [comment]
Thanks, looks good apart from future-proofing.

**Comment by brodybits** [comment]
I pushed some changes but please do not merge it yet! It fails when I try to run it on node 0.10.40, will investigate. Closing for now, maybe will raise a new PR.

File: test/cpp/wrappedobjectfactory.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+NAN_MODULE_INIT(Init) {
+  MyObject::Init(target);
+  Set(target
+    , New<v8::String>("newFactoryObjectInstance").ToLocalChecked()
+    , GetFunction(New<v8::FunctionTemplate>(MyObject::NewInstance)).ToLocalChecked()
+  );
+}
+
+NODE_MODULE(objectwraphandle, Init)
```
Comment (kkoopa) [inline test/cpp/wrappedobjectfactory.cpp]: They failed because this module is not called objectwraphandle.

**Comment by brodybits** [comment]
The problems should now be fixed, I tried it on Node 0.10.40 and 4.1.1 (was OK on Node 0.12.7 and 4.0.0 in the past). I added another change where the wrapped object can return another wrapped object (I need this functionality for my own project), and rebased on the latest master changes.

**Comment by kkoopa** [comment]
Looks good. Test failure on travis seems to be a network error downloading something.

**Comment by brodybits** [comment]
Travis aargh @#$%

Thanks for looking at this.

**Comment by brodybits** [comment]
I pushed a couple more changes:

1. Small fix to `test/js/wrappedobjectfactory-test.js`:
```diff
--- a/test/js/wrappedobjectfactory-test.js
+++ b/test/js/wrappedobjectfactory-test.js
@@ -10,7 +10,7 @@ const test     = require('tap').test
     , testRoot = require('path').resolve(__dirname, '..')
     , bindings = require('bindings')({ module_root: testRoot, bindings: 'wrappedobjectfactory' });
 
-test('objectwraphandle', function (t) {
+test('wrappedobjectfactory', function (t) {
   t.plan(4);
 
   var obj = bindings.newFactoryObjectInstance(123);
```

2. Also fix `test/cpp/objectwraphandle.cpp` to move constructor object into a private static inline function

**Comment by kkoopa** [comment]
Merged as 4392aa9df745960d553a47e61ae58d8bb811753c bff133ab771eef56315bb2f498a0d95777246f1f 3a4a0ae15bcf35e177d35087063146151c56449f

**Comment by brodybits** [comment]
Thanks for merging my changes!

**Pull request closed without merging**